deepspeed/runtime/zero/stage3.py

```diff
@@ -1087,8 +1071,6 @@ def __reduce_and_partition_ipg_grads(self, safe_mode: bool = False) -> None:
                 raise RuntimeError(f"{param.grad.numel()} != {param.ds_numel} Cannot reduce scatter "
                                    f"gradients whose size is not same as the params")
 
-        self.params_in_ipg_bucket.sort(key=lambda p: p.ds_id)
```

Comment:
The check in line 1083 will still pass assuming the order is identical between ranks.

Reply:
My understanding is that the `sort()` was added because there is no guarantee or identical order across the ranks. Also, the `safe_mode:` test condition is added as debugging mode.

Reply:
Params are added to params_in_ipg_bucket list in the same order they are added to the ipg_bucket_flat_buffer in __add_grad_to_ipg_bucket, the order of backward hooks. The backward hooks order is identical between ranks.

Reply:
ok. sorry for misunderstanding.
Yes, removing the sort is necessary since inside partition_grads called in line 1092 (new code) params_in_ipg_bucket and grad_partitions are assumed to have the same order of param/grad pairs.
However, with the new code, grad_bucket will contain grad tensors in the order they appear in __ipg_bucket_flat_buffer and can't be reordered to match the sorted param list.